### PR TITLE
Separate run participants graphql participants move

### PR DESCRIPTION
**Pull request by trilopin**
#### Summary
This PR introduces the new graphql run mutations to add and remove participants to a run.

```graphql
	addRunParticipants(runID: String!, userIDs: [String!]!): String!
	removeRunParticipants(runID: String!, userIDs: [String!]!): String!
```

RootResolver has been split on Run & Playbook embedded objects in order to organize a bit the growing graphql codebase. 1/5 on this move, I'm 100% open to discussion. This is the main reason the PR has many changes.

RunResolver completed with missing fields. Not really necessary for this PR but one step forward more GQL usage.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-playbooks/issues/1402

#### Checklist
- [ ] ~~Telemetry updated~~
- [ ] ~~Gated by experimental feature flag~~
- [X] Unit tests updated

### CODE REVIEW

**Review by crspeller — CHANGES_REQUESTED** [review]
Awesome! Just one question about the refactor.

**Review by gbochora — COMMENTED** [review]
Great work @trilopin! I've left some questions and comments.

File: server/api/graphql_run.go [hunk 1]
```diff
@@ -32,3 +91,183 @@ func (r *RunResolver) IsFavorite(ctx context.Context) (bool, error) {
...
+
+	if err := c.playbookRunService.AddParticipants(args.RunID, args.UserIDs); err != nil {
+		return "", errors.Wrap(err, "failed to add participant from run")
+	}
+
+	for _, userID := range args.UserIDs {
+		if err := c.playbookRunService.Follow(args.RunID, userID); err != nil {
+			return "", errors.Wrap(err, "failed to make participant to unfollow run")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/api/graphql_run.go:243]: Currently, becoming a participant implies following, and removing from participants implies unfollowing. We feel this behavior won't be like that forever, that's the main reason to keep them separated at service/store and just call both cases from handler/resolver.

File: server/api/graphql_run.go [hunk 2]
```diff
@@ -32,3 +91,183 @@ func (r *RunResolver) IsFavorite(ctx context.Context) (bool, error) {
...
+
+	if err := c.playbookRunService.RemoveParticipants(args.RunID, args.UserIDs); err != nil {
+		return "", errors.Wrap(err, "failed to remove participant from run")
+	}
+
+	for _, userID := range args.UserIDs {
+		if err := c.playbookRunService.Unfollow(args.RunID, userID); err != nil {
+			return "", errors.Wrap(err, "failed to make participant to unfollow run")
+		}
+	}
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/api/graphql_run.go:270]: [same case](https://github.com/mattermost/mattermost-plugin-playbooks/pull/1424#discussion_r952257865)

File: server/app/playbook_run_service.go [hunk 3]
```diff
@@ -2196,7 +2198,13 @@ func (s *PlaybookRunServiceImpl) UserHasLeftChannel(userID, channelID, actorID s
 		s.logger.Errorf("faied to remove participant that left channel for run '%s', user '%s'; error: %s", playbookRunID, user.Id, err.Error())
 	}
 
-	_ = s.sendPlaybookRunToClient(playbookRunID)
+	if err := s.Unfollow(playbookRunID, user.Id); err != nil {
+		s.logger.Errorf("failed to make participant to unfollow the run '%s', user '%s'; error: %s", playbookRunID, user.Id, err.Error())
+	}
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/app/playbook_run_service.go:2203]: At this point, leaving the channel implies leaving the run. Additionally, leaving the run implies unfollowing.

That will change after introducing actions, and here we will have a boolean flag that enable/disable this behavior.

File: webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx [hunk 4]
```diff
@@ -40,53 +38,41 @@ interface Props {
...
-        } else if (!channel) {
-            addToast(formatMessage({defaultMessage: 'Your request wasn\'t successful.'}), ToastType.Failure);
-            return;
-        }
 
-        // if channel is not null, join the channel
-        await dispatch(joinChannel(currentUserId, playbookRun.team_id, playbookRun.channel_id, playbookRunMetadata.channel_name));
+        addToRun()
+            .then(() => addToast(formatMessage({defaultMessage: 'You\'ve joined this run.'}), ToastType.Success))
+            .catch(() => addToast(formatMessage({defaultMessage: 'It wasn\'t possible to join the run'}), ToastType.Failure));
```
[the remark this answers is not in the document]
Comment (trilopin) [reply webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx:74]: Not sure if I understand you. This is about joining a run, not a channel.

[the remark this answers is not in the document]
Comment (gbochora) [reply webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx:74]: Yep, I can see it's pretty unclear : ))
If we have a public run with a private channel, adding a user to run members also means adding to the channel. 

https://github.com/mattermost/mattermost-plugin-playbooks/blob/10b892f5c771a9d7a1f0cbfc6f621b09012d31ca/server/app/playbook_run_service.go#L2817

So by clicking on the `Participate` button user will become a member of the private channel.

[the remark this answers is not in the document]
Comment (trilopin) [reply webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx:74]: Oh, I got it. You are right

From webapp, we don't add to channel anymore. However, we are temporarily adding the user to channel when it's added to run. That shouldn't happen if the is participate-flow with private channel.

Good catch!

File: server/api/graphql_run.go [hunk 5]
```diff
@@ -32,3 +91,183 @@ func (r *RunResolver) IsFavorite(ctx context.Context) (bool, error) {
...
+	RunID   string
+	UserIDs []string
+}) (string, error) {
+	c, err := getContext(ctx)
+	if err != nil {
+		return "", err
+	}
+	userID := c.r.Header.Get("Mattermost-User-ID")
+
+	if err := c.permissions.RunView(userID, args.RunID); err != nil {
```
[the remark this answers is not in the document]
Comment (trilopin) [reply server/api/graphql_run.go]: Thanks @gbochora, two points here:
- any user with access to a run can add other user (and this internally relies on channel members)
- since it relies on channel members... it needs to be migrated to rely on run members

https://github.com/mattermost/mattermost-plugin-playbooks/issues/1432

**Comment by trilopin** [comment]
@gbochora @crspeller Thank you for your comments

Several changes:
- refactor refactored :) Naming improved and rootresolvers separated from resolvers. Still 2/5, open to more changes if you have ideas.
- e2e tests for graphql run root resolvers 
- handled edge case (+test) when participating in a public run with private channel


Created new issue to handle permissions https://github.com/mattermost/mattermost-plugin-playbooks/issues/1432

File: webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply webapp/src/components/backstage/playbook_runs/playbook_run/header.tsx:74]: I thought we had decided this case was OK? Since the user has access to the run and the action is configured to automatically add them to the channel. 
I am fine with keeping it this way until a explicit action is configurable though.

**Review by crspeller — APPROVED** [review]

File: server/app/playbook_run_service.go [hunk 6]
```diff
@@ -2823,19 +2823,29 @@ func (s *PlaybookRunServiceImpl) AddParticipants(playbookRunID string, userIDs [
...
 
-func (s *PlaybookRunServiceImpl) participateActions(playbookRun *PlaybookRun, userID string) {
+func (s *PlaybookRunServiceImpl) participateActions(playbookRun *PlaybookRun, userID string, requesterUserID string) {
 	// Don't do anything if the user is a channel member
 	member, _ := s.pluginAPI.Channel.GetMember(playbookRun.ChannelID, userID)
 	if member != nil {
 		return
 	}
 
+	channel, err := s.pluginAPI.Channel.Get(playbookRun.ChannelID)
```
Comment (gbochora) [inline server/app/playbook_run_service.go:2839]: Optional: Would it be better to move this call outside the `participateActions` function and pass information as an argument? We will avoid multiple calls for each user.

**Review by gbochora — APPROVED** [review]
Looks good! Thanks, @trilopin 🥳

File: server/app/playbook_run_service.go (hunk 6 above)
Comment (trilopin) [reply server/app/playbook_run_service.go:2839]: > Optional: Would it be better to move this call outside the `participateActions` function and pass information as an argument? We will avoid multiple calls for each user.

but the channel is not queried before, is it?

Comment (trilopin) [reply server/app/playbook_run_service.go:2839]: nice catch @gbochora, same find inside a loop for every user... 🗡️
